### PR TITLE
Voting With The Majority

### LINKED ISSUE #4117: ETH1Data majority preference voting
Follow up from #4002.

Part of the honest validator guide recommends voting with the majority eth1data votes. 
See the write up for more context.

https://hackmd.io/@prestonvanloon/H1rfsG5iH

### PR DESCRIPTION
**Pull request by rkapka**
**What type of PR is this?**

Feature

**What does this PR do? Why is it needed?**

The aim of this PR is to enhance the proposer with a new voting algorithm as described here: https://hackmd.io/@prestonvanloon/H1rfsG5iH#Voting-With-The-Majority

**Which issues(s) does this PR fix?**

Fixes #4117

**Other notes for review**

~There are some `TODO`s in the code because I need clarification on a couple of things. I would rather not guess how things should work by myself.~

### CODE REVIEW

File: shared/params/mainnet_config.go [hunk 1]
```diff
@@ -123,6 +123,7 @@ var mainnetBeaconConfig = &BeaconChainConfig{
 	MaxPeersToSync:            15,
 	SlotsPerArchivedPoint:     2048,
 	GenesisCountdownInterval:  time.Minute,
+	HeadOfVotingPeriodLength:  32,
```
Comment (rkapka) [inline shared/params/mainnet_config.go:126]: From the linked document:

> In mainnet configuration, the head of the voting period is the first 32 slots of the period `sqrt(1024) = 32`

File: shared/params/minimal_config.go [hunk 2]
```diff
@@ -76,5 +76,7 @@ func MinimalSpecConfig() *BeaconChainConfig {
 
 	minimalConfig.DepositContractTreeDepth = 32
 	minimalConfig.FarFutureEpoch = 1<<64 - 1
+	minimalConfig.HeadOfVotingPeriodLength = 6
```
Comment (rkapka) [inline shared/params/minimal_config.go:79]: `SLOTS_PER_ETH1_VOTING_PERIOD = 32` in the minimal config, so I chose the number which is closes to `sqrt(32)`.

File: beacon-chain/rpc/validator/proposer.go [hunk 3]
```diff
@@ -194,6 +206,128 @@ func (vs *Server) eth1Data(ctx context.Context, slot uint64) (*ethpb.Eth1Data, e
...
 
+// eth1DataMajorityVote determines the appropriate eth1data for a block proposal using an extended
+// simple voting algorithm - voting with the majority. The algorithm for this method is as follows:
+//  - Determine if we are in the "head" of the voting period. If yes, use the simple voting algorithm for proposal.
+//  - Determine the timestamp for the start slot for the current eth1 voting period.
+//  - Determine the timestamp for the start slot for the previous eth1 voting period.
+//  - Determine the most recent eth1 block before each timestamp.
+//  - Subtract the current period's eth1block.number by ETH1_FOLLOW_DISTANCE to determine the voting upper bound.
+//  - Subtract the previous period's eth1block.number by ETH1_FOLLOW_DISTANCE to determine the voting lower bound.
+//  - Filter out votes on unknown blocks and blocks which are outside of the range determined by the lower and upper bounds.
```
Comment (rkapka) [inline beacon-chain/rpc/validator/proposer.go:217]: The `unknown blocks` part is something that is not in the linked document, but I added this condition because I feel it is important. Because Eth1Data does not include the block height, it might happen that the block which has the most votes is outside of the range of blocks that we want to vote on. Because of this we cannot safely vote on it, so we need to discard it.

File: beacon-chain/rpc/validator/proposer.go [hunk 4]
```diff
@@ -194,6 +206,128 @@ func (vs *Server) eth1Data(ctx context.Context, slot uint64) (*ethpb.Eth1Data, e
...
+
+				voteCount[i].votes++
+				newVote = false
+				break
+			}
+		}
+
+		if newVote {
+			voteCount = append(voteCount, eth1DataAggregatedVote{data: singleVote, votes: 1})
+		}
```
Comment (rkapka) [inline beacon-chain/rpc/validator/proposer.go:315]: I kind of dislike this algorithm. I wanted to use a simple `map[ethpb.Eth1Data]int` or something to keep track of the votes, but the compiler complained about equality not being defined on the type. I started looking for a solution, but everything was either discouraged from using outside of tests (e.g. `DeepEqual`) or I would need to implement custom equality somehow. All in all I decided to quit looking for now and just implement an obscure and naive algorithm.

Any ideas on how to make this better are welcome!

File: shared/params/mainnet_config.go (hunk 1 above)
Comment (terencechain) [inline shared/params/mainnet_config.go:126]: `HeadOfVotingPeriodLength` can easily be confused as one of the beacon chain config. 
I suggest we add `eth1` to the name

File: beacon-chain/rpc/validator/proposer.go [hunk 5]
```diff
@@ -194,6 +206,128 @@ func (vs *Server) eth1Data(ctx context.Context, slot uint64) (*ethpb.Eth1Data, e
...
+//  - Determine the timestamp for the start slot for the current eth1 voting period.
+//  - Determine the timestamp for the start slot for the previous eth1 voting period.
+//  - Determine the most recent eth1 block before each timestamp.
+//  - Subtract the current period's eth1block.number by ETH1_FOLLOW_DISTANCE to determine the voting upper bound.
+//  - Subtract the previous period's eth1block.number by ETH1_FOLLOW_DISTANCE to determine the voting lower bound.
+//  - Filter out votes on unknown blocks and blocks which are outside of the range determined by the lower and upper bounds.
+//  - If no blocks are left after filtering, use the current period's most recent eth1 block for proposal.
+//  - Determine the vote with the highest count. Prefer the vote with the highest eth1 block height in the event of a tie.
+//  - This vote's block is the eth1block to use for the block proposal.
+// TODO: "Genesis" period and previous period
```
Comment (rkapka) [inline beacon-chain/rpc/validator/proposer.go:221]: The algorithm checks for a few things from the previous voting period. It will also be executed during the first `SLOTS_PER_ETH1_VOTING_PERIOD` proposals, during which there is no previous period. This means that this case has to be handled separately, right? I just want to make sure I am thinking about this one correctly.

Comment (rkapka) [inline beacon-chain/rpc/validator/proposer.go:221]: The algorithm checks for a few things from the previous voting period. It will also be executed during the first `SLOTS_PER_ETH1_VOTING_PERIOD` proposals, during which there is no previous period. This means that this case has to be handled separately, right? I just want to make sure I am thinking about this one correctly.

File: beacon-chain/rpc/validator/proposer.go [hunk 6]
```diff
@@ -194,6 +206,128 @@ func (vs *Server) eth1Data(ctx context.Context, slot uint64) (*ethpb.Eth1Data, e
...
+		return vs.randomETH1DataVote(ctx)
+	}
+	eth1DataNotification = false
+
+	headOfVotingPeriod :=
+		slot%(params.BeaconConfig().EpochsPerEth1VotingPeriod*params.BeaconConfig().SlotsPerEpoch) <
+			params.BeaconConfig().HeadOfVotingPeriodLength
+	if headOfVotingPeriod {
+		// TODO: Is it OK that `context.WithTimeout` will be invoked again?
+		return vs.eth1Data(ctx, slot)
```
Comment (rkapka) [inline beacon-chain/rpc/validator/proposer.go:240]: I am invoking `eth1Data()` from within `eth1DataMajorityVote()` and both functions begin with `context.WithTimeout(ctx, eth1dataTimeout)`. Is it a problem that the context function will be invoked twice?

File: beacon-chain/rpc/validator/proposer.go [hunk 7]
```diff
@@ -194,6 +206,128 @@ func (vs *Server) eth1Data(ctx context.Context, slot uint64) (*ethpb.Eth1Data, e
...
+//  - Determine the most recent eth1 block before each timestamp.
+//  - Subtract the current period's eth1block.number by ETH1_FOLLOW_DISTANCE to determine the voting upper bound.
+//  - Subtract the previous period's eth1block.number by ETH1_FOLLOW_DISTANCE to determine the voting lower bound.
+//  - Filter out votes on unknown blocks and blocks which are outside of the range determined by the lower and upper bounds.
+//  - If no blocks are left after filtering, use the current period's most recent eth1 block for proposal.
+//  - Determine the vote with the highest count. Prefer the vote with the highest eth1 block height in the event of a tie.
+//  - This vote's block is the eth1block to use for the block proposal.
+// TODO: "Genesis" period and previous period
+// TODO: Feature flag
+func (vs *Server) eth1DataMajorityVote(ctx context.Context, slot uint64) (*ethpb.Eth1Data, error) {
```
Comment (terencechain) [inline beacon-chain/rpc/validator/proposer.go:223]: `eth1DataMajorityVote` is long, i suggest we refactor some into smaller functions. 
For one example, the aggregation logic can be it's own function

File: beacon-chain/rpc/validator/proposer.go [hunk 8]
```diff
@@ -194,6 +206,128 @@ func (vs *Server) eth1Data(ctx context.Context, slot uint64) (*ethpb.Eth1Data, e
...
+	var inRangeVotes []eth1DataSingleVote
+	for _, eth1Data := range headState.Eth1DataVotes() {
+		ok, height, err := vs.BlockFetcher.BlockExists(ctx, bytesutil.ToBytes32(eth1Data.BlockHash))
+		if err != nil {
+			log.WithError(err).Warning("Could not fetch eth1data height for received eth1data vote")
+		}
+		// TODO: Both inclusive?
+		if ok && lastValidBlockNumber.Cmp(height) > -1 && firstValidBlockNumber.Cmp(height) < 1 {
+			inRangeVotes = append(inRangeVotes, eth1DataSingleVote{eth1Data: *eth1Data, blockHeight: height})
+		}
```
Comment (rkapka) [inline beacon-chain/rpc/validator/proposer.go:287]: The linked document describes the algorithm in such a way that both the upper and lower bound blocks are included in the possible result set. Because of this the same block can be potentially voted for in two succeeding periods (the upper bound for the previous period becomes the lower bound for the next period). Are we OK with this or should one of the bounds be exclusive?

File: beacon-chain/rpc/validator/proposer.go (hunk 7 above)
Comment (nisdas) [reply beacon-chain/rpc/validator/proposer.go:223]: +1, this function is doing 10 different things. It can be refactored into a few smaller functions

Comment (nisdas) [reply beacon-chain/rpc/validator/proposer.go:223]: +1 this method is hard to parse, its doing too many things and should be broken up in smaller functions

File: beacon-chain/rpc/validator/proposer.go (hunk 6 above)
Comment (nisdas) [reply beacon-chain/rpc/validator/proposer.go:240]: yes that is fine

File: beacon-chain/rpc/validator/proposer.go (hunk 4 above)
Comment (nisdas) [reply beacon-chain/rpc/validator/proposer.go:315]: `reflect.DeepEqual`, `ssz.DeepEqual`

File: shared/params/mainnet_config.go (hunk 1 above)
Comment (nisdas) [reply shared/params/mainnet_config.go:126]: +1 it conflicts with a well known param.

Comment (prestonvanloon) [reply shared/params/mainnet_config.go:126]: Please add a comment
```suggestion
	HeadOfVotingPeriodLength:  32, // sqrt(1024)
```

File: beacon-chain/rpc/validator/proposer.go (hunk 3 above)
Comment (prestonvanloon) [reply beacon-chain/rpc/validator/proposer.go:217]: :+1: good idea

**Comment by prestonvanloon** [comment]
I've just noticed that this bit of logic was removed from the spec earlier this year in https://github.com/ethereum/eth2.0-specs/pull/1553.

When should look into why the spec is recommending to immediately vote with the majority, even if there is only a single vote. 
Here's a bit more on the discussion: https://github.com/ethereum/eth2.0-specs/issues/1537

File: shared/params/mainnet_config.go (hunk 1 above)
Comment (rkapka) [reply shared/params/mainnet_config.go:126]: This parameter has been removed. See Preston's comment for more details.

File: shared/featureconfig/flags.go [hunk 9]
```diff
@@ -156,12 +156,16 @@ var (
...
 	initSyncVerbose = &cli.BoolFlag{
 		Name:  "init-sync-verbose",
-		Usage: "Enable logging every processed block during initial syncing. ",
+		Usage: "Enable logging every processed block during initial syncing.",
 	}
 	enableFinalizedDepositsCache = &cli.BoolFlag{
 		Name:  "enable-finalized-deposits-cache",
 		Usage: "Enables utilization of cached finalized deposits",
 	}
+	enableEth1DataMajorityVote = &cli.BoolFlag{
```
Comment (prestonvanloon) [inline shared/featureconfig/flags.go:165]: Please add this to devModeFlags

Comment (rkapka) [reply shared/featureconfig/flags.go:165]: Added in #6768
